pkg/build/buildkit/k8s_autodiscovery.go

```diff
@@ -0,0 +1,338 @@
+	w, err := cs.CoreV1().Pods(ns).Watch(ctx, metav1.ListOptions{
+		LabelSelector: labelSelector,
+		Watch:         true,
+	})
+	if err != nil {
+		return err
+	}
+
+	defer w.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+
+		case e := <-w.ResultChan():
+			if e.Type != watch.Added && e.Type != watch.Modified {
+				continue
+			}
+
+			pod := e.Object.(*corev1.Pod)
+			if isPodReady(pod) {
+				pods <- pod
+			}
+
+			klog.V(4).Infof("Pod %s/%s is not ready yet", pod.Namespace, pod.Name)
+		}
+	}
+}
+
+func acquireLeaseForPod(ctx context.Context, cs *kubernetes.Clientset, ch chan<- *corev1.Pod, pod *corev1.Pod, opts KubernertesDiscoveryOptions) {
+	podname := os.Getenv("POD_NAME")
+	if podname == "" {
+		podname, _ = os.Hostname()
+	}
+
+	klog.V(4).Infof("Attempting to acquire the lease for pod %s/%s under holder name %q...", pod.Namespace, pod.Name, podname)
+
+	leaderelection.RunOrDie(ctx, leaderelection.LeaderElectionConfig{
+		Lock: &resourcelock.LeaseLock{
+			LeaseMeta: metav1.ObjectMeta{
+				Name:      fmt.Sprintf("%s-%s", strings.TrimRight(opts.LeasePrefix, "-"), pod.Name),
+				Namespace: pod.Namespace,
+			},
+			Client: cs.CoordinationV1(),
+			LockConfig: resourcelock.ResourceLockConfig{
+				Identity: podname,
+			},
+		},
+		ReleaseOnCancel: true,
+		LeaseDuration:   5 * time.Second,
+		RenewDeadline:   2 * time.Second,
+		RetryPeriod:     500 * time.Millisecond,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(c context.Context) {
+				select {
+				case ch <- pod:
+					klog.V(4).Infof("Selected BuildKit pod: %s/%s", pod.Namespace, pod.Name)
+
+				case <-ctx.Done():
+					klog.V(4).Infof("Received context cancelation: %s/%s", pod.Namespace, pod.Name)
+				}
+			},
+			OnStoppedLeading: noopFunc,
+		},
+	})
+
+	klog.V(4).Infof("Shutting off the lease for %s/%s pod", pod.Namespace, pod.Name)
+}
+
+func setTsuruAppLabelOnBuildKitPod(ctx context.Context, cs *kubernetes.Clientset, pod, ns, app string) error {
+	patch, err := json.Marshal([]any{
+		map[string]any{
+			"op":    "replace",
+			"path":  fmt.Sprintf("/metadata/labels/%s", normalizeAppLabelForJSONPatch(TsuruAppNameLabelKey)),
+			"value": app,
+		},
+		map[string]any{
+			"op":    "replace",
+			"path":  fmt.Sprintf("/metadata/labels/%s", normalizeAppLabelForJSONPatch(TsuruIsBuildLabelKey)),
+			"value": strconv.FormatBool(true),
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	_, err = cs.CoreV1().Pods(ns).Patch(ctx, pod, types.JSONPatchType, patch, metav1.PatchOptions{})
+	return err
+}
+
+func unsetTsuruAppLabelOnBuildKitPod(ctx context.Context, cs *kubernetes.Clientset, pod, ns string) error {
+	patch, err := json.Marshal([]any{
+		map[string]any{
+			"op":   "remove",
+			"path": fmt.Sprintf("/metadata/labels/%s", normalizeAppLabelForJSONPatch(TsuruAppNameLabelKey)),
+		},
+		map[string]any{
+			"op":   "remove",
+			"path": fmt.Sprintf("/metadata/labels/%s", normalizeAppLabelForJSONPatch(TsuruIsBuildLabelKey)),
+		},
```

Comment:
we could also add a annotation to inform the time of last build.

Reply:
Can you take another look? :eyes:

Reply:
<3